engine/classes/Elgg/Search/SearchService.php

```diff
@@ -0,0 +1,307 @@
+				$options = $this->hooks->trigger('search:options', "$entity_type:$entity_subtype", $params, $params);
+			}
+
+		} else {
+			$fields = $this->hooks->trigger('search:fields', "$search_type", $params, $fields);
+			if (!isset($params['fields'])) {
+				$params['fields'] = $fields;
+			} else {
+				// only allow known fields
+				$params['fields'] = array_intersect($fields, $params['fields']);
+			}
+
+			if ($this->hooks->hasHandler('search:options', "$search_type")) {
+				$options = $this->hooks->trigger('search:options', "$search_type", $params, $params);
+			}
+		}
+
+		if (!$options) {
+			return false;
+		}
+		
+		return call_user_func($getter, $options);
+	}
+
+	/**
+	 * Prepare default entity search clauses
+	 *
+	 * @param string $type          Entity type
+	 * @param string $query         Search query
+	 * @param array  $fields        Metadata and attribute fields to search in
+	 * @param string $sort          Sorting field
+	 * @param string $order         Sorting order (asc|desc)
+	 * @param bool   $partial_match Allow partial matches (e.g. find 'elgg' for 'el' search)
+	 * @param bool   $tokenize      Break down search query into token (e.g. find 'elgg has been released' for 'elgg release')
+	 * @return array|false
+	 */
+	public function getSearchClauses($type, $query, $fields, $sort, $order, $partial_match = true, $tokenize = true) {
+
+		switch ($type) {
+			case 'user' :
+				$table = 'users_entity';
+				$secondary_attributes = array_keys(\ElggUser::getExternalAttributes());
+				break;
+
+			case 'group' :
+				$table = 'groups_entity';
+				$secondary_attributes = array_keys(\ElggGroup::getExternalAttributes());
+				break;
+
+			case 'object' :
+				$table = 'objects_entity';
+				$secondary_attributes = array_keys(\ElggObject::getExternalAttributes());
+				break;
+
+			default :
+				return false;
+		}
+		
+		$primary_attributes = array_keys(\ElggEntity::getPrimaryAttributes());
+
+		$clauses = [
+			'wheres' => [],
+			'joins' => [],
+		];
+
+		$clauses['joins'][] = "JOIN {$this->db->prefix}{$table} secondary ON e.guid = secondary.guid";
+
+		foreach ($fields as $field) {
+			if (in_array($field, $primary_attributes)) {
+				// we do not allow search in primary attributes
+			} else if (in_array($field, $secondary_attributes)) {
+				$attribute_fields[] = $field;
+			} else {
+				$metadata_fields[] = $field;
+			}
+		}
+
+		if (!empty($attribute_fields)) {
+			$attributes_where = $this->getWhereSQL('secondary', $attribute_fields, $query, $partial_match, $tokenize);
+		}
+
+		if (!empty($metadata_fields)) {
+			$clauses['joins'][] = "JOIN {$this->db->prefix}metadata md_search on e.guid = md_search.entity_guid";
+
+			// get the where clauses for the md names
+			// can't use egef_metadata() because the n_table join comes too late.
+			$md_clauses = _elgg_entities_get_metastrings_options('metadata', [
+				'metadata_names' => $metadata_fields,
+				// avoid notices
```

Comment:
the fact that without the `null` values you will get notices... this feels silly